File: problems/basic/641/solution.cpp
// Accepted: 21ms (67.23%), 21.30MB (5.35%)

class MyCircularDeque {
    int internalSize;
    vector<int> array;
    int front;
    int rear;

    int getNextIndex(int i) {
        return (i+1) % internalSize;
    }

    int getPrevIndex(int i) {
        return (i+internalSize-1) % internalSize;
    }

public:
    MyCircularDeque(int k) {
        internalSize = k+1;
        array.resize(internalSize);
        front = 0;
        rear = 0;
    }

    bool insertFront(int value) {
        if (isFull()) {
            return false;
        }

        array[front] = value;
        front = getNextIndex(front);
        return true;
    }

    bool insertLast(int value) {
        if (isFull()) {
            return false;
        }

        rear = getPrevIndex(rear);
        array[rear] = value;
        return true;
    }

    bool deleteFront() {
        if (isEmpty()) {
            return false;
        }

        front = getPrevIndex(front);
        return true;
    }

    bool deleteLast() {
        if (isEmpty()) {
            return false;
        }

        rear = getNextIndex(rear);
        return true;

    }

    int getFront() {
        if (isEmpty()) {
            return -1;
        }

        return array[getPrevIndex(front)];
    }

    int getRear() {
        if (isEmpty()) {
            return -1;
        }

        return array[rear];
    }

    bool isEmpty() {
        return front == rear;
    }

    bool isFull() {
        return getNextIndex(front) == rear;
    }
};

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * MyCircularDeque* obj = new MyCircularDeque(k);
 * bool param_1 = obj->insertFront(value);
 * bool param_2 = obj->insertLast(value);
 * bool param_3 = obj->deleteFront();
 * bool param_4 = obj->deleteLast();
 * int param_5 = obj->getFront();
 * int param_6 = obj->getRear();
 * bool param_7 = obj->isEmpty();
 * bool param_8 = obj->isFull();
 */

File: problems/basic/641/solution.ts
// Accepted: 109ms (66.67%), 60.58MB (9.52%)

class MyCircularDeque {
    private internalSize: number;
    private array: number[];
    private front: number;
    private rear: number;

    constructor(k: number) {
        this.internalSize = k+1;
        this.array = new Array(k+1).fill(0);
        this.front = 0;
        this.rear = 0;
    }

    private getNextIndex(i: number) {
        return (i+1) % this.internalSize;
    }

    private getPrevIndex(i: number) {
        return (i+this.internalSize-1) % this.internalSize;
    }

    insertFront(value: number): boolean {
        if (this.isFull()) {
            return false;
        }

        this.array[this.front] = value;
        this.front = this.getNextIndex(this.front);
        return true;
    }

    insertLast(value: number): boolean {
        if (this.isFull()) {
            return false;
        }

        this.rear = this.getPrevIndex(this.rear);
        this.array[this.rear] = value;
        return true;
    }

    deleteFront(): boolean {
        if (this.isEmpty()) {
            return false;
        }

        this.front = this.getPrevIndex(this.front);
        return true;
    }

    deleteLast(): boolean {
        if (this.isEmpty()) {
            return false;
        }

        this.rear = this.getNextIndex(this.rear);
        return true;
    }

    getFront(): number {
        if (this.isEmpty()) {
            return -1;
        }

        const front = this.getPrevIndex(this.front);
        return this.array[front];
    }

    getRear(): number {
        if (this.isEmpty()) {
            return -1;
        }

        return this.array[this.rear];
    }

    isEmpty(): boolean {
        return this.front === this.rear;
    }

    isFull(): boolean {
        return this.getNextIndex(this.front) === this.rear;
    }
}

/**
 * Your MyCircularDeque object will be instantiated and called as such:
 * var obj = new MyCircularDeque(k)
 * var param_1 = obj.insertFront(value)
 * var param_2 = obj.insertLast(value)
 * var param_3 = obj.deleteFront()
 * var param_4 = obj.deleteLast()
 * var param_5 = obj.getFront()
 * var param_6 = obj.getRear()
 * var param_7 = obj.isEmpty()
 * var param_8 = obj.isFull()
 */
